src/redprl/metalanguage/dynamics.fun

```diff
@@ -0,0 +1,163 @@
+functor MetalanguageDynamics (ML : METALANGUAGE) = 
+struct
+  datatype hole = HOLE
+  datatype ('a, 'b) closure = <: of 'a * ('b, 'b) closure ML.Ctx.dict
+  
+  structure Rules = Refiner (Signature)
+  structure J = RedPrlSequent and CJ = RedPrlCategoricalJudgment
+
+  type mlterm = ML.mlterm_
+  type mlscope = (ML.mlvar, mlterm) ML.mlscope
+
+  datatype continuation =
+     FUN of hole * mlterm
+   | ARG of mlscope * hole
+   | LET of ML.mlvar * hole * mlterm
+   | PAIR0 of hole * mlterm
+   | PAIR1 of mlterm * hole
+   | FST of hole
+   | SND of hole
+   | EACH of {tactics: mlterm list, done: Lcf.jdg Lcf.Tl.telescope, metaenv: Tm.metaenv}
+   | SPLICE of {tactics: mlterm list, var: Lcf.L.var, done: Lcf.jdg Lcf.Tl.telescope, remaining: Lcf.jdg Lcf.Tl.telescope, metaenv: Tm.metaenv}
+
+  type stack = (continuation, mlterm) closure list
+
+  type names = int -> Tm.symbol
+  datatype state = LOCAL of Lcf.jdg | GLOBAL of Lcf.jdg Lcf.state
+  datatype control = ## of (state * names) * (mlterm, mlterm) closure
+  datatype machine = |> of control * stack | <| of control * stack
```

Comment:
Currently the machine regards the proof state as "state", and executes scripts and programs against it. This seems the most obvious thing to do, but it would be better if we could somehow disentangle things so that the machine is actually producing in the end a (semantic) tactic. This would make it easier to program the behavior of builtin tacticals in SML, without having to be able to encode their functionality in RedML.

I'm not certain if I can do what I described above, but I'm going to think about it.